File: exercises_book/cli/word_frequency_finder.mjs
import PromptSync from "prompt-sync"; // Import prompt-sync for user input (not used in this script)
const prompt = PromptSync({ sigint: true }); // Initialize prompt-sync

import fs from "node:fs/promises"; // Import fs/promises for async file operations
import { time } from "node:console";

const filePath = "../files/macbeth.txt"; // Path to the input text file

// Asynchronously read the contents of the input file
async function readFile(filePath) {
  try {
    let data = await fs.readFile(filePath, "utf8");
    return data;
  } catch (error) {
    console.log(error);
  }
}

// Process the text: remove punctuation, split into words, and count frequencies
function processWords(data) {
  let words_freq = {};
  data = data.replace(/[.,!?]/g, ""); // Remove basic punctuation
  data = data
    .split(/\r?\n/) // Split by newlines
    .join(" ")      // Join lines into a single string
    .split(" ")     // Split by spaces to get words
    .filter((word) => word !== ""); // Remove empty strings
  // Count frequency of each word (case-insensitive)
  data.map((word) => (
    words_freq[word.toLowerCase()] = words_freq[word.toLowerCase()] 
      ? words_freq[word.toLowerCase()] + 1 
      : 1
  ));
  return words_freq;
}

// Sort the words by frequency in descending order
function sortWords(words_freq) {
  let sorted_words = Object.fromEntries(
    Object.entries(words_freq).sort((a, b) => b[1] - a[1])
  );
  return sorted_words;
}

// Print the histogram: word followed by asterisks representing frequency
function print(sorted_words){
   try{
    for(const [key, value] of Object.entries(sorted_words)){
        console.log(`${key.padEnd(10, ' ')}: ${'*'.repeat(value)}`);
    }
   }catch(error){
    console.log(error);
   }
}

// Main function to coordinate reading, processing, sorting, and printing
async function main() {
  try {
    console.time("run-time");
    const words = await readFile(filePath);      // Read the file
    const processed = processWords(words);       // Process and count words
    const sorted = sortWords(processed);         // Sort by frequency
    print(sorted);
    console.timeEnd("run-time");                              // Print the histogram
  } catch (error) {
    console.log(error);
  }
}

main(); // Start

File: exercises_book/cli/word_frequency_finder.py
import re
import matplotlib.pyplot as plt
file_path = '../files/macbeth.txt'  # Path to the input text file
import time

# Reading the file and returning its content as a single string
def read_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read()
        content = content.replace('\n', ' ')  # Replace newlines with spaces
    return content

# Remove punctuation from the text and convert all words to lowercase
def remove_punctuation(text):
   edited_text = re.sub(r"[.,?!]", "", text)  # Remove specified punctuation
   edited_text = edited_text.split(' ')       # Split text into words
   edited_text = list(filter(None,edited_text))  # Remove empty strings
   edited_text = [word.lower() for word in edited_text]  # Convert to lowercase
   return edited_text

# Count the frequency of each word and return a sorted dictionary (descending by frequency)
def create_sort_dict(array):
    words_freq = {}
    for word in array:
        if word in words_freq:
            words_freq[word] += 1
        else:
            words_freq[word] = 1
    # Sort the dictionary by frequency in descending order
    sorted_words_freq = dict(sorted(words_freq.items(), key=lambda item: item[1], reverse=True))
    return sorted_words_freq

# Print the histogram: word followed by asterisks representing frequency
def show_result(dict):
    for key, value in dict.items():
        print(f'{key.ljust(10, " ")} : {"*"*value}')

# Plot a horizontal bar chart of the top N word frequencies using matplotlib
def plot_histogram(freq_dict, top_n=10):
    words = list(freq_dict.keys())[:top_n]      # Get top N words
    counts = list(freq_dict.values())[:top_n]   # Get their counts
    plt.barh(words[::-1], counts[::-1], color="skyblue")  # Plot in reverse for highest at top
    plt.xlabel("Frequency")
    plt.title(f"Top {top_n} Word Frequencies")
    plt.tight_layout()
    plt.show()

# Main function to coordinate reading, processing, displaying, and timing
def main():
    start_time = time.time()  # Start timing
    words = read_file(file_path)              # Read the file
    words = remove_punctuation(words)         # Clean and split into words
    words_freq = create_sort_dict(words)      # Count and sort frequencies
    show_result(words_freq)                   # Print the histogram to the console
    end_time = time.time() 
    elapsed_time = (end_time - start_time) * 1000  # Calculate elapsed time in ms
    print(f"Execution time: {elapsed_time:.2f} ms")
    plot_histogram(words_freq)                # Show graphical bar chart

main()  # Run the program